pyvista/core/utilities/geometric_sources.py

```diff
@@ -0,0 +1,252 @@
+"""Provides an easy way of generating several geometric sources.
+
+as well as some pure-python helpers.
+
+"""
+from typing import Sequence
+
+from pyvista.core import _vtk_core as _vtk
+from pyvista.core.utilities.misc import no_new_attr
+
+from .helpers import wrap
+
+
+@no_new_attr
+class ConeSource(_vtk.vtkConeSource):
+    """Cone source algorithm class.
+
+    Parameters
+    ----------
+    center : sequence[float], default: (0.0, 0.0, 0.0)
+        Center in ``[x, y, z]``. Axis of the cone passes through this
+        point.
+
+    direction : sequence[float], default: (1.0, 0.0, 0.0)
+        Direction vector in ``[x, y, z]``. Orientation vector of the
+        cone.
+
+    height : float, default: 1.0
+        Height along the cone in its specified direction.
+
+    radius : float, optional
+        Base radius of the cone.
+
+    capping : bool, optional
+        Enable or disable the capping the base of the cone with a
+        polygon.
+
+    angle : float, optional
+        The angle in degrees between the axis of the cone and a
+        generatrix.
+
+    resolution : int, default: 6
+        Number of facets used to represent the cone.
+
+    Examples
+    --------
+    Create a default ConeSource.
+
+    >>> import pyvista
+    >>> source = pyvista.ConeSource()
+    >>> source.output.plot(show_edges=True, line_width=5)
+    """
+
+    def __init__(
+        self,
+        center=(0.0, 0.0, 0.0),
+        direction=(1.0, 0.0, 0.0),
+        height=1.0,
+        radius=None,
+        capping=True,
+        angle=None,
+        resolution=6,
+    ):
+        """Initialize the cone source class."""
+        super().__init__()
+        self.center = center
+        self.direction = direction
+        self.height = height
+        self.capping = capping
+        if angle is not None and radius is not None:
+            raise ValueError("Both radius and angle specified. They are mutually exclusive.")
+        elif angle is not None and radius is None:
+            self.angle = angle
+        elif angle is None and radius is not None:
+            self.radius = radius
+        elif angle is not None and radius is None:
```

Comment:
Oh. This is not correct!
```suggestion
        elif angle is not None and radius is not None:
```

Reply:
Fix in 66eefe8b80edb2c5743eafe0af9d4ba9f6c3cdfe . I need coffee.